compare/meson.build

```diff
@@ -0,0 +1,19 @@
+compare_src = files(
+    'compare.c',
+    'image.c',
+)
+
+libass_compare = executable(
+    'libass_compare',
+    compare_src,
+    install: false,
+    include_directories: incs,
+    dependencies: deps + test_deps,
+    link_with: libass_for_tools,
+)
+
+art_samples = get_option('art-samples')
+if art_samples != ''
+    dir = join_paths(art_samples, 'regression')
+    test('art-compare', files(join_paths(dir, 'run-all.sh'))[0], workdir: dir, args: [libass_compare.full_path()], depends: [libass_compare])
```

Comment:
I would suggest using native binaries to run tests. It is more portable. As it is defined currently, meson doesn't know that the binary is executed, only path is passed to the script. Therefore it won't add varaibles to environment like PATH, ASAN_OPTIONS and most imprtantly when cross compiling it wouldn't know to use exe_wrapper (like wine or qemu) if needed. All this makes running tests portable a nightmare, while it would be already handled by Meson.

The easiest and canonical way would be to loop the files in meson and declare test for each of them, instead of using `run_all.sh` helper.

Reply:
Further, looking at the CI output, it seems that this causes Meson to swallow the output of our test scripts and just print a single “OK” for the whole lot of tests. Which is certainly better than nothing—and at the end of the day we can run the tests manually to find the failing test—but not ideal. Although maybe it prints the output if the tests should fail?

Actually, does it even properly detect failures?

But anyway, on the other hand, the suggested solution seems to amount to a full rewrite of our test infrastructure in Meson, which also isn’t ideal. On the third hand, this might be easier than it sounds, because some of that code is too generalized/configurable for basic CI and could perhaps be simplified during this rewrite.

Reply:
By default meson will redirect the output into `meson-logs`, but this can be configured with e.g. `-v`

Reply:
> Although maybe it prints the output if the tests should fail?

It logs everything in test-log and can be enable to stdout with verbose flag. Though if it passes shorter output generally make it easier to ready, when there is no need for more info.

> But anyway, on the other hand, the suggested solution seems to amount to a full rewrite of our test infrastructure in Meson

I don't see it that way, I had a glance at run_all.sh script. And it is essentiality aggregator of testcases. So what already Meson does and sure it would be duplicating some logic, so running script like that is fine, but as you noticed it bundles all tests into one, which may not be ideal. Though it is not critical to change it, I feel like it would simplify the flow for Meson.

Reply:
I already have a step that prints the full log if tests fail.

Reply:
But anyway, as far as I'm aware, meson doesn't support globbing or other methods of listing the contents of a directory, so this suggestion is not actually possible without hardcoding the list of tests (from `libass/libass-tests`) within `libass/libass`. I suppose it might be possible to `run_command('ls')` and then parse the output? But that'd mean pulling in a list of tests at configure-time, which would be brittle (i.e. if you added a test case, it wouldn't get pulled in until you re-ran meson), and I don't think it'd provide any actual *benefit* here.

Reply:
> But anyway, as far as I'm aware, meson doesn't support globbing or other methods of listing the contents of a directory

And that's a good thing. But you can list dir or whatever you need with a script. https://mesonbuild.com/FAQ.html#but-i-really-want-to-use-wildcards

Reply:
Also my point is mostly, that if MSVC Windows support is something that is a real target it should work without third-party tools like sh.exe (above other things) from MSYS2. I know it is currently not feasible, but this is generally something to keep in mind.